### PR TITLE
Nflog support MVP

### PR DESCRIPTION
**Pull request by vorner**
Hello

This adds the support for basic use of the nflog protocol. There are still some parts missing, but I'd prefer to have a feedback about it sooner than later and I tried to make sure all the missing pieces can be added without breaking API.

I've also tried to structure it in a way that makes it possible/natural to add all the other netfilter protocols over time (NFQUEUE, CONNTRACK).

There are, however, few things I'm not entirely happy about or don't know how to handle properly. I'll probably open separate issues to discuss them one by one, if there's some nice solution for them. But I'll still list my worries here first.

* When receiving messages over the socket, I have to specify into what type I want them to deserialize. But this might depend on the type. This can be seen on the added example, sometimes I get a message with type 3 ‒ which I think corresponds to NLMSG_DONE. I'm not sure if I should make it part of the new `NetfilterMsg` enum, but that would still not help much, because it would still try to deserialize the rest as `LogPacket`. This doesn't currently fail, but is kind of wrong. It would be great if I could somehow specify a nl-type => parse type relation. Unlike `NlAttr<_, Vec<u8>>`, the Nlmsghdr<_, Vec<u8>>` doesn't have methods to parse as one or another type.
* Some of the structs/messages are ever sent in only one direction. So it makes no sense to provide eg. `serialize` for the `LogPacket`. For now, I've put `unimplemented` in them, but it kind of doesn't feel right. Would it make sense to split the Nl trait into reading and writing halves?
* The crate uses the `Vec` type quite generously. For example, when parsing a message, it allocates a Vec as the receive buffer, then it allocates a Vec for all the attributes and, as they have diffferent types, each one is a smaller Vec... As the Vec contains a heap allocation, it might turn somewhat expensive. This probably doesn't matter for things like manipulating the routing table, but with the NFLOG, the number of packets received per second can be quite huge. It would be great if there was a way to use the library with smaller number of allocations (some of my ideas could get down to 0 per packet).
* The `impl_*` macros don't want to let me put documentation onto the individual variants. I've tried to poke at them a little bit, but I've managed to only break existing code, so I've given up for now.

Anyway, if I'm missing anything important here, or if there's any way to improve the code or interface, I'm happy to hear it. I'd then add some fixup commits and rebuild the branch before merging.

### CODE REVIEW

**Comment by vorner** [comment]
Linking here #37, the macros...

**Comment by vorner** [comment]
CI can't compile libc :-(.

**Comment by vorner** [comment]
Note to self. I've just discovered methods like `u32::to_be`. This means all the `htons` are not necessary and I'll get rid of them in a fixup some time soonish.

**Comment by jbaublitz** [comment]
Restart CI here too.

**Comment by jbaublitz** [comment]
CI is fixed.

File: src/consts/netfilter.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+//! be added over time.
+
+use super::{NlAttrType, NlType};
+
+impl_var_trait! {
+    /// Attributes inside a netfilter log packet message.
+    ///
+    /// These are send by the kernel and describe a logged packet.
+    NfLogAttr, u16, NlAttrType,
+    PacketHdr => 1,
```
Comment (jbaublitz) [inline src/consts/netfilter.rs]: My guess is that these constants don't show up in libc. If we can get them into libc, I'd much prefer that over hard coding constants. We do this elsewhere, but I'd have to see if the libc crate would be open to accepting a PR for that.

File: src/netfilter.rs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+#[derive(Copy, Clone, Debug)]
+struct Timestamp {
+    secs: u64,
+    usecs: u64,
+}
+
+impl Nl for Timestamp {
+    fn serialize(&self, m: &mut StreamWriteBuffer) -> Result<(), SerError> {
+        u64::to_be(self.secs).serialize(m)?;
```
Comment (jbaublitz) [inline src/netfilter.rs]: Are you positive these are all in big endian? If you are assuming that because netlink is a network protocol, netlink always uses the endianness of the host system, whether big or little, as it is exclusively a local protocol. If you have an example of this protocol specifically being big endian values, can you provide that evidence? [This comment](https://chromium.googlesource.com/native_client/linux-headers-for-nacl/+/2dc04f8190a54defc0d59e693fa6cff3e8a916a9/include/linux/netlink.h#140) was how I originally discovered this as netlink will explicitly let you know when it's network byte order as opposed to host byte order.

File: src/socket.rs [hunk 3]
```diff
@@ -336,9 +336,9 @@ impl NlSocket {
             Some(ref mut b) => Nlmsghdr::deserialize(b)?,
             None => unreachable!(),
         };
-        if self.pid.is_none() {
+        if self.pid.is_none() && msg.nl_pid != 0 {
```
Comment (jbaublitz) [inline src/socket.rs]: Please move the socket changes into a separate PR. I'm happy to merge those before this one. I agree that PID tracking is very broken and I'm looking to fix that in v0.5.0 but am happy to turn this off for now in another PR.

Comment (vorner) [reply src/socket.rs]: OK, I've submitted the commit as #57. It's the same one, so it should disappear from the diff here once that one is merged into master.

File: src/netfilter.rs (hunk 2 above)
Comment (vorner) [reply src/netfilter.rs]: I'm positive. First, the type in the headers are marked as `__be32`.

Second, I did it without these conversions first and it didn't work ‒ the values for timestamp, for example, usually produced overflows when converting to SystemTime. When I do the conversion, it produces sane values.

File: src/consts/netfilter.rs (hunk 1 above)
Comment (vorner) [reply src/consts/netfilter.rs]: I'll open an issue there. Should we wait with this pull request until they decide, or go ahead and update it once a new release is made on their side?

File: src/netfilter.rs (hunk 2 above)
Comment (jbaublitz) [reply src/netfilter.rs]: Excellent. This looks good then. I know it can get a little tedious providing evidence so thank you for bearing with me.

File: src/consts/netfilter.rs (hunk 1 above)
Comment (jbaublitz) [reply src/consts/netfilter.rs]: I'll merge this instead of waiting on that because that may take a bit to figure out and they don't always release all that quickly. Once I do merge this, I'm going to open an issue that, if you're willing, I'll assign to you to convert those over to libc constants if that gets merged eventually. Sound good? Also, can you link to the issue you opened here so I can follow up on it? Thanks!

File: src/socket.rs (hunk 3 above)
Comment (jbaublitz) [reply src/socket.rs]: Excellent thank you! I want to think through that one a little more before merging, hence the separate PR. Once we figure that out, I'll ask you to rebase onto master so that you pull in whatever we decide on and then I'll merge this one.

File: src/consts/netfilter.rs (hunk 1 above)
Comment (vorner) [reply src/consts/netfilter.rs]: Sounds good.

https://github.com/rust-lang/libc/issues/1562

File: src/netfilter.rs (hunk 2 above)
Comment (vorner) [reply src/netfilter.rs]: That's fine. I take a thorough review as a seal of quality of the library :-).

**Comment by jbaublitz** [comment]
Hi @vorner, thanks for your other contribution. I've merged that PR so can you rebase onto master and force push here? I'll do one final review before merging.
